File: apps/api/app/schemas/email.py
"""Email inbox API schemas."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class EmailMessageResponse(BaseModel):
    """A synchronized email message."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    thread_id: UUID

    provider_message_id: str
    provider_thread_id: str

    internet_message_id: str | None = None
    in_reply_to: str | None = None
    references_header: str | None = None

    sender_name: str | None = None
    sender_email: str | None = None

    recipient_emails: list[str] = Field(default_factory=list)
    cc_emails: list[str] = Field(default_factory=list)

    subject: str | None = None
    text_body: str | None = None
    html_body: str | None = None

    received_at: datetime | None = None
    sent_at: datetime | None = None

    direction: str
    unread: bool
    has_attachments: bool

    created_at: datetime
    updated_at: datetime


class EmailThreadSummary(BaseModel):
    """Summary information for one inbox thread."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    provider_thread_id: str

    subject: str | None = None
    snippet: str | None = None

    participant_emails: list[str] = Field(default_factory=list)

    last_message_at: datetime | None = None
    unread: bool

    message_count: int = 0

    created_at: datetime
    updated_at: datetime


class EmailThreadListResponse(BaseModel):
    """Paginated thread-list response."""

    items: list[EmailThreadSummary]

    total: int
    page: int
    page_size: int
    total_pages: int


class EmailThreadDetail(BaseModel):
    """A thread with all synchronized messages."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    provider_thread_id: str

    subject: str | None = None
    snippet: str | None = None

    participant_emails: list[str] = Field(default_factory=list)

    last_message_at: datetime | None = None
    unread: bool

    created_at: datetime
    updated_at: datetime

    messages: list[EmailMessageResponse] = Field(
        default_factory=list
    )


class EmailThreadSummaryResponse(BaseModel):
    """AI-generated summary for one email thread."""

    thread_id: UUID
    summary: str = Field(
        min_length=1,
        max_length=4000,
    )
    message_count: int = Field(
        ge=0,
    )

class ReplySuggestion(BaseModel):
    """One AI-generated reply suggestion."""

    subject: str = Field(
        min_length=1,
        max_length=255,
    )

    body: str = Field(
        min_length=1,
        max_length=10000,
    )


class ReplySuggestionsResponse(BaseModel):
    """Collection of reply suggestions for one email thread."""

    thread_id: UUID

    suggestions: list[ReplySuggestion] = Field(
        default_factory=list,
    )

class DraftReplyRequest(BaseModel):
    """Request to create a Gmail draft reply."""

    subject: str = Field(
        min_length=1,
        max_length=255,
    )
    body: str = Field(
        min_length=1,
        max_length=10000,
    )


class DraftReplyResponse(BaseModel):
    """Created Gmail draft reply."""

    thread_id: UUID
    draft_id: str
    message_id: str | None = None
    status: str

class SalesOpportunityResponse(BaseModel):
    """Sales-opportunity classification for one email thread."""

    thread_id: UUID
    classification: str = Field(
        min_length=1,
        max_length=50,
    )
    confidence: float = Field(
        ge=0.0,
        le=1.0,
    )
    reason: str = Field(
        min_length=1,
        max_length=1000,
    )

File: web/src/types/email.ts
export interface GmailSyncResponse {
  fetched: number
  created: number
  updated: number
  skipped: number
  failed: number
}

export interface EmailMessage {
  id: string
  thread_id: string
  provider_message_id: string
  provider_thread_id: string
  internet_message_id: string | null
  in_reply_to: string | null
  references_header: string | null
  sender_name: string | null
  sender_email: string | null
  recipient_emails: string[]
  cc_emails: string[]
  subject: string | null
  text_body: string | null
  html_body: string | null
  received_at: string | null
  sent_at: string | null
  direction: string
  unread: boolean
  has_attachments: boolean
  created_at: string
  updated_at: string
}

export interface EmailThreadSummary {
  id: string
  provider_thread_id: string
  subject: string | null
  snippet: string | null
  participant_emails: string[]
  last_message_at: string | null
  unread: boolean
  message_count: number
  created_at: string
  updated_at: string
}

export interface EmailThreadListResponse {
  items: EmailThreadSummary[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface EmailThreadDetail {
  id: string
  provider_thread_id: string
  subject: string | null
  snippet: string | null
  participant_emails: string[]
  last_message_at: string | null
  unread: boolean
  created_at: string
  updated_at: string
  messages: EmailMessage[]
}

export interface ThreadListParams {
  page?: number
  page_size?: number
  search?: string
  sender?: string
  unread?: boolean
}

export interface EmailThreadSummaryResponse {
  thread_id: string
  summary: string
  message_count: number
}

export interface ReplySuggestion {
  subject: string
  body: string
}

export interface ReplySuggestionsResponse {
  thread_id: string
  suggestions: ReplySuggestion[]
}

export interface DraftReplyRequest {
  subject: string
  body: string
}

export interface DraftReplyResponse {
  thread_id: string
  draft_id: string
  message_id: string | null
  status: string
}

export interface SalesOpportunityResponse {
  thread_id: string
  classification:
    | 'sales_opportunity'
    | 'non_sales'
    | 'unknown'
  confidence: number
  reason: string
}

export interface QuotationExtractionResponse {
  thread_id: string
  product: string | null
  quantity: number | null
  pricing_requested: boolean
  availability_requested: boolean
  delivery_requested: boolean
  payment_terms_requested: boolean
  confidence: number
}

export interface QuotationPreviewResponse {
  thread_id: string

  customer_id: string
  customer_name: string
  customer_email: string | null

  product_id: string
  product_name: string

  quantity: string
  unit: string
  unit_price: string
  currency: string
  tax_rate: string

  subtotal: string
  tax_amount: string
  total_amount: string
}

export interface CreateQuotationFromThreadResponse {
  thread_id: string
  quotation_id: string
  quotation_number: string
  status: string
}

export type QuotationStatus =
  | 'draft'
  | 'reviewed'
  | 'approved'
  | 'sent'
  | 'accepted'
  | 'rejected'
  | 'expired'
  | 'cancelled'

export interface QuotationStatusUpdateRequest {
  status: QuotationStatus
}

export interface QuotationStatusUpdateResponse {
  id: string
  quotation_number: string
  status: QuotationStatus
}

export interface QuotationDraftRequest {
  quotation_id: string
}

export interface QuotationDraftResponse {
  thread_id: string
  quotation_id: string
  quotation_number: string

  draft_id: string
  message_id: string | null

  recipient: string
  attachment_filename: string

  status: string
}